File: src/main/java/scalastudy/extend/FieldConf.scala
package scalastudy.extend

/**
  * Created by shuqin on 17/4/10.
  * Scala 实现枚举; 由于 Enumeration 不支持枚举含有方法字段，因此采用样例对象模拟实现
  */
sealed class FieldConf(name:String, title:String, able: (Person)=>String) {
  def getTitle = title
  def getAble = able
}

object FieldConf {

  def apply(name: String): FieldConf = {
    name match {
      case "Id" => Id
      case "Name" => Name
      case "Able" => Able
      case _ => Unknown
    }
  }
}

case object Id extends FieldConf("Id", "编号", p => p.getId)
case object Name extends FieldConf("Name", "姓名", p => p.getName)
case object Able extends FieldConf("Able", "能力", p => p.getAble)
case object Unknown extends FieldConf("Unknown", "未知", p => "")

File: src/main/java/zzz/study/function/FieldConf.java
package zzz.study.function;

import java.util.function.Function;

/**
 * Created by shuqin on 17/3/30.
 */
public enum FieldConf {

    Id("Id", "编号", Person::getId),
    Name("Name", "姓名", Person::getName),
    Able("Able", "能力", Person::able);

    private String name;
    private String title;
    private Function<Person, String> able;

    FieldConf(String name, String title, Function<Person,String> able) {
        this.name = name;
        this.title = title;
        this.able = able;
    }

    public String getName() {
        return name;
    }

    public String getTitle() {
        return title;
    }

    public Function<Person, String> getAble() {
        return able;
    }
}
